File: Plugins/AudioPlayer/src/ios_AudioPlayer/AudioPlayer/MetaDataReader.swift
//
//  MetaDataReader.swift
//  AudioPlayer
//
//  Created by Gary Griswold on 8/7/17.
//  Copyright © 2017 ShortSands. All rights reserved.
//

import AWS

class MetaDataReader {
    
    var oldTestament: TOCAudioBible?
    var newTestament: TOCAudioBible?
    var metaDataVerse: TOCAudioChapter?
    
    init() {
        self.oldTestament = nil
        self.newTestament = nil
    }
    
    deinit {
        print("***** Deinit MetaDataReader *****")
    }

    func read(versionCode: String, complete: @escaping (_ oldTest:TOCAudioBible?, _ newTest:TOCAudioBible?) -> Void) {
        let db = Sqlite3()
        let query = "SELECT a.damId, a.collectionCode, a.mediaType, a.dbpLanguageCode, a.dbpVersionCode" +
                " FROM audio a, audioVersion v" +
                " WHERE a.dbpLanguageCode = v.dbpLanguageCode" +
                " AND a.dbpVersionCode = v.dbpVersionCode" +
                " AND v.versionCode = ?" +
                " ORDER BY mediaType ASC, collectionCode ASC"
                // mediaType sequence Drama, NonDrama
                // collectionCode sequence NT, ON, OT
        do {
            try db.open(dbPath: "Versions.db", copyIfAbsent: true)
            defer { db.close() }
            try db.queryV1(sql: query, values: [versionCode], complete: { resultSet in
                print("LENGTH \(resultSet.count)")
                var oldTestRow: [String?]? = nil
                var newTestRow: [String?]? = nil
                for row in resultSet {
                    // Because of the sort sequence, the following logic prefers Drama over Non-Drama
                    // Because of the sequenc of IF's, it prefers OT and NT over ON
                    let collectionCode = row[1]!
                    if newTestRow == nil && collectionCode == "NT" {
                        newTestRow = row
                    }
                    if newTestRow == nil && collectionCode == "ON" {
                        newTestRow = row
                    }
                    if oldTestRow == nil && collectionCode == "OT" {
                        oldTestRow = row
                    }
                    if oldTestRow == nil && collectionCode == "ON" {
                        oldTestRow = row
                    }
                }
                if let oldRow = oldTestRow {
                    self.oldTestament = TOCAudioBible(database: db, mediaSource: "FCBH", dbRow: oldRow)
                }
                if let newRow = newTestRow {
                    self.newTestament = TOCAudioBible(database: db, mediaSource: "FCBH", dbRow: newRow)
                }
                complete(self.oldTestament, self.newTestament)
            })
        } catch let err {
            print("ERROR \(Sqlite3.errorDescription(error: err))")
            complete(nil, nil)
        }
    }
    /**
    * This function will only return results after read has been called.
    */
    func findBook(bookId: String) -> TOCAudioBook? {
        var result: TOCAudioBook? = nil
        if self.oldTestament != nil {
            result = self.oldTestament!.booksById[bookId]
        }
        if self.newTestament != nil {
            result = self.newTestament!.booksById[bookId]
        }
        return result
    }
    
    func readVerseAudio(damid: String, sequence: String, bookId: String, chapter: String,
                        complete: @escaping (_ audioVerse: TOCAudioChapter?) -> Void) {
        self.metaDataVerse = nil
        let s3Bucket = damid.lowercased() + ".shortsands.com"
        let s3Key = "Verse_" + sequence + "_" + bookId + ".json"
        AwsS3Cache.shared.readData(s3Bucket: s3Bucket,
               s3Key: s3Key,
               expireInterval: 604800, // 1 week in seconds
               getComplete: { data in
                    if let verses = data {
                        let result = self.parseJsonDictionary(json: verses)
                        if let chapterNum = Int(chapter) {
                            let chapterStr = String(chapterNum)
                            if let dictionary = result?[chapterStr] as? NSDictionary {
                                self.metaDataVerse = TOCAudioChapter(chapterDictionary: dictionary)
                            } else {
                                print("ERROR: Chapter not found in verse meta data.")
                            }
                        } else {
                            print("ERROR: Chapter was invalid number.")
                        }
                    } else {
                        print("ERROR: There was no data retrieved.")
                    }
                    complete(self.metaDataVerse)
            })
    }
    
    private func parseJsonDictionary(json: Data) -> NSDictionary? {
        do {
            let result = try JSONSerialization.jsonObject(with: json, options:.allowFragments) as? NSDictionary
            return result
        } catch let jsonError {
            print("Error parsing Meta Data json \(jsonError)")
            return nil
        }
    }
}

File: Plugins/AudioPlayer/src/android_AudioPlayer/app/src/main/java/com/shortsands/audioplayer/MetaDataReader.java
package com.shortsands.audioplayer;

/**
 * Created by garygriswold on 8/30/17.
 */
import android.content.Context;
import android.util.Log;
import java.util.HashMap;
import org.json.JSONException;
import org.json.JSONArray;
import org.json.JSONObject;

class MetaDataReader {

    private static final String TAG = "MetaDataReader";

    private final Context context;
    private final HashMap<String, TOCAudioBible> metaData;
    private TOCAudioChapter metaDataVerse;
    private CompletionHandler readCompletion;
    private CompletionHandler readVerseCompletion;

    MetaDataReader(Context context) {
        this.context = context;
        this.metaData = new HashMap<String, TOCAudioBible>();
    }

    void read(String languageCode, String mediaType, CompletionHandler completion) {
        this.readCompletion = completion;
        AWSS3Cache cache = new AWSS3Cache(this.context);
        String s3Bucket = "audio-us-west-2-shortsands";
        String s3Key = languageCode + "_" + mediaType + ".json";
        int expireInterval = 604800; // seconds in 1 week
        ReadResponseHandler handler = new ReadResponseHandler();
        cache.readText(s3Bucket, s3Key, expireInterval, handler);
    }

    class ReadResponseHandler implements CompletionHandler {

        public void completed(Object result) {
            Log.d(TAG, "***** Inside Completed in MetaDataReader");
            JSONArray json = parseJson(result);
            if (json != null) {
                Log.d(TAG, "JSON PARSED " + json.toString());
                for (int i=0; i<json.length(); i++) {
                    try {
                        JSONObject item = json.getJSONObject(i);
                        TOCAudioBible metaItem = new TOCAudioBible("FCBH", item);
                        Log.d(TAG, "TOCAudioBible item: " + metaItem.toString());
                        metaData.put(metaItem.damId, metaItem);
                    } catch(JSONException je) {
                        Log.e(TAG, "Could not parse Audio Meta Data " + je.toString());
                        readCompletion.failed(je);
                    }
                }
                readCompletion.completed(metaData);
            } else {
                Log.d(TAG, "Not parsable JSON");
                readCompletion.failed(new RuntimeException("Could not parse JSON"));
            }
        }
        public void failed(Throwable exception) {
            Log.e(TAG, "Exception in MetaDataReader.read " + exception.toString());
            readCompletion.failed(exception);
        }
    }

    void readVerseAudio(String damid, String sequence, String bookId, String chapter,
                               CompletionHandler completion) {
        this.readVerseCompletion = completion;
        AWSS3Cache cache = new AWSS3Cache(this.context);
        String s3Bucket = "audio-us-west-2-shortsands";
        String s3Key = damid + "_" + sequence + "_" + bookId + "_" + chapter + "_verse.json";
        int expireInterval = 604800; // seconds in 1 week
        ReadVerseResponseHandler handler = new ReadVerseResponseHandler();
        cache.readText(s3Bucket, s3Key, expireInterval, handler);
    }

    class ReadVerseResponseHandler implements CompletionHandler {

        public void completed(Object result) {
            JSONArray json = parseJson(result);
            if (json != null) {
                metaDataVerse = new TOCAudioChapter(json);
                readVerseCompletion.completed(metaDataVerse);
            } else {
                readVerseCompletion.failed(new RuntimeException("Failed to parse JSON"));
            }
        }
        public void failed(Throwable exception) {
            Log.e(TAG, "Exception in MetaDataReader.readVerseAudio " + exception.toString());
            readVerseCompletion.failed(exception);
        }
    }

    private JSONArray parseJson(Object data) {
        if (data != null) {
            if (data instanceof String) {
                try {
                    JSONArray json = new JSONArray((String) data);
                    return json;
                } catch(JSONException exc) {
                    Log.e(TAG, "Error parsing Meta Data json " + exc.toString());
                    return null;
                }
            } else {
                Log.e(TAG, "Downloaded Meta Data is not String.");
                return null;
            }
        } else {
            Log.e(TAG, "Download Meta Data Error.");
            return null;
        }
    }
}
